external/dhcpd/dhcpd_lwip.c

```diff
@@ -92,15 +87,85 @@
 #define DHCP_SERVER_PORT         67
 #define DHCP_CLIENT_PORT         68
 
-/****************************************************************************
- * Private Types
- ****************************************************************************/
-/****************************************************************************
- * Private Data
- ****************************************************************************/
+static pthread_t g_dhcpd_tid = -1;
+static int g_dhcpd_term = 0;
+
+struct dhcp_join_data {
+	char *intf;
+	dhcp_sta_joined fn;
+};
+typedef struct dhcp_join_data dhcp_join_data_s;
+
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+#define DHCPD_MQ_NAME "dhcpd_queue"
+#define DHCPD_MQ_LEN 11
+#define DHCPD_MQ_MAX_LEN 20
+
+void *_dhcpd_join_handler(void *arg)
+{
+	dhcp_join_data_s *data = (dhcp_join_data_s *)arg;
+	int sockfd = 0;
+	struct req_lwip_data req;
+	int ret = ERROR;
+	struct mq_attr attr;
+	attr.mq_maxmsg = DHCPD_MQ_MAX_LEN;
+	attr.mq_msgsize = DHCPD_MQ_LEN;
+	attr.mq_flags = 0;
+	attr.mq_curmsgs = 0;
+
+	mqd_t md = mq_open(DHCPD_MQ_NAME, O_RDWR | O_NONBLOCK |O_CREAT, 0666, &attr);
+	if (md == (mqd_t)ERROR) {
+		ndbg("Failed to open mq\n");
+		goto go_out;
+	}
+
+	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		ndbg("socket() failed with errno: %d\n", errno);
+		goto go_out;
+	}
+
+	memset(&req, 0, sizeof(req));
+	req.type = DHCPDSTART;
+	req.host_name = data->intf;
+
+	ret = ioctl(sockfd, SIOCLWIP, (unsigned long)&req);
+	if (ret == ERROR) {
+		ndbg("ioctl() failed with errno: %d\n", errno);
+		goto go_out;
+	}
+
+	while (!g_dhcpd_term) {
```

Comment:
You'd better classify code between related things for all the time (in this case handle socket first and then mq)

Also if it is possible, timed receive is more stable than mq_receive

Reply:
I totally agree with you. However, here there is a timing issue. After running LwIP DHCP through ioctl, there might be sending msg operation from LwIP DHCP, before mq is opened. Thanks for the comment.